resources/Materials/TestSuite/Utilities/envmap.osl

```diff
@@ -1,11 +1,33 @@
-shader envmap(float Kb = 1, string filename = "")
+void calculateLodBias(float lodBias, output float lod)
 {
-   vector dir = normalize(I);
-   float radial = atan2(-dir[2], dir[0]);
-   float nradial = acos(dir[1]);
-   float r = 0.5 * sin(nradial * 0.5);
-   float tu = 0.5 + r * cos(radial);
-   float tv = 0.5 - r * sin(radial);
-   color c = texture(filename, tu, tv);
-   Ci = Kb * c * background();
+    lod = lodBias < 0.25 ? sqrt(lodBias) : 0.5*lodBias + 0.375;
+}
+
+void latlong_map_lookup(vector dir, string filename, float lodBias, output color result)
+{
+   vector transformedDir = rotate(dir, M_PI_2, vector(0, 0, -1));
+   float M_PI_INV = 1.0/M_PI;
+   float latitude = -asin(transformedDir[0]) * M_PI_INV + 0.5;
+   latitude = clamp(latitude, 0.01, 0.99);
+   float longitude = atan2(transformedDir[1], -transformedDir[2]) * M_PI_INV * 0.5 + 0.5;
+   if (lodBias > 0)
+   {
+      float lod;
+      calculateLodBias(lodBias, lod);
+      result = texture(filename, longitude, latitude, "blur", lod);
+   }
+   else
+   {
+      result = texture(filename, longitude, latitude);
+   }
+}
+
+shader envmap(float Kb = 1, string radiance_filename = "", string irradiance_filename = "", float specular_roughness = 0)
+{
+   vector R = reflect (I, N);
+   color c;
+   color c2;
+   latlong_map_lookup(R, radiance_filename, specular_roughness, c);
+   latlong_map_lookup(N, irradiance_filename, 0, c2);
+   Ci = Kb * (c + c2) * background();
```

Comment:
Using a lod bias based on roughness is a hack for HW rendering, and we should not use this for OSL / offline rendering. For offline rendering (path/ray tracing) we can rely on the given "I" direction being the right direction to sample, and do point sampling on it, since the ray direction "I" for indirect rays is generated based on the BSDF being sampled. So from the shader's point of view we get a correct and importance sampled environment "for free".

So you only need to worry about using the right projection here. Keep the lat-long calculation but remove the lod blur. And don't use the reflected direction R, you should use "I" directly. You have to imagine "I" is the ray direction both for camera rays and for secondary indirect rays. So for the secondary case "I" has already been reflected of the surface in a direction given by the BSDF lobe.

There is also no need to use separate radiance and irradiance textures. Only a single environment texture should be used, and the BSDF lobes will handle blurring this correctly and automatically.

Reply:
Thanks for the tips!